src/connection.rs

```diff
@@ -0,0 +1,188 @@
+        let mut payload = vec![0; length as usize];
+        self.stream.read_exact(&mut payload)?;
+
+        // Evaluate payload
+        let message = MessageType::from_raw(id, &payload);
+
+        // Handle special messages
+        match &message {
+            // Save the current session
+            ServerSync(data) => self.session = Some(data.get_session()),
+
+            // Keep a list of all users
+            UserState(data) => {
+                // Replace or append user
+                if let Some(index) = self
+                    .users
+                    .iter()
+                    .position(|u| u.get_session() == data.get_session())
+                {
+                    // If there is a new channel id we update it
+                    if data.has_channel_id() {
+                        self.users[index].set_channel_id(data.get_channel_id());
+                    }
+                } else {
+                    // Add user
+                    self.users.push(*data.clone());
+                }
+            }
+            // Ignore other messages
+            _ => {}
+        };
+
+        Ok(message)
+    }
+
+    pub fn send(&mut self, message: &MessageType) {
+        // Create a stream to output the message
+        let mut output_stream = CodedOutputStream::new(&mut self.stream);
+
+        // Send bytes
+        output_stream.write_raw_bytes(&message.to_raw()).unwrap();
+        output_stream.flush().unwrap();
+    }
+
+    pub fn send_version(&mut self) {
+        // Send version message
+        let message = mumble::Version::new();
+        self.send(&Version(message));
+    }
+
+    pub fn send_authentication(&mut self, username: &str, password: &str) {
+        // Send authentication message
+        let mut message = mumble::Authenticate::new();
+        message.set_username(username.to_string());
+        message.set_password(password.to_string());
+        self.send(&Authenticate(message));
+    }
+
+    pub fn send_text_message(&mut self, text: &str) {
+        // Send text message
+        match self.session {
+            Some(s) => {
+                let mut message = mumble::TextMessage::new();
+                message.set_actor(s);
+                message.set_channel_id(vec![0]);
+                message.set_message(text.to_string());
+                self.send(&TextMessage(message));
+            }
+            None => println!("Connection not established. (No session)"),
+        }
+    }
+
+    pub fn change_channel(&mut self, channel_id: u32) {
+        // Send channel change message
+        if let Some(s) = self.session {
+            let mut message = mumble::UserState::new();
+            message.set_session(s);
+            message.set_actor(s);
+            message.set_channel_id(channel_id);
+            self.send(&UserState(Box::new(message)));
+        } else {
+            println!("Connection not established. (No session)");
+        }
+    }
+
+    pub fn keep_alive(&mut self) {
```

Comment:
I don't think I like the polling approach here. E.g. I think `examples/connect.rs` would get disconnected if there is no valid message from the server for 30s. The ping sounds like something the connection should handle internally, at least by default.

Thoughts? @Prior99 @svenstaro 

Either way, `SystemTime` is not monotonic and not the right tool here. `Instant` should be better.

Reply:
I switched to `Instant`, but we still have polling.